File: core-java/src/main/java/com/kakao/actionbase/core/java/state/SpecialStateValue.java
package com.kakao.actionbase.core.java.state;

public enum SpecialStateValue {
  DELETED("__DELETED__"),
  UNSET("__UNSET__");

  final String stateValue;

  SpecialStateValue(String code) {
    this.stateValue = code;
  }

  public String code() {
    return stateValue;
  }

  public static boolean isSpecialStateValue(Object code) {
    if (code instanceof String) {
      return code.equals(DELETED.code()) || code.equals(UNSET.code());
    } else {
      return false;
    }
  }

  public static SpecialStateValue getSpecialStateValue(String code) {
    if (DELETED.code().equals(code)) {
      return DELETED;
    } else if (UNSET.code().equals(code)) {
      return UNSET;
    } else {
      throw new IllegalArgumentException("Invalid special state value: " + code);
    }
  }
}

File: core/src/main/kotlin/com/kakao/actionbase/core/state/SpecialStateValue.kt
package com.kakao.actionbase.core.state

enum class SpecialStateValue(
    private val stateValue: String,
) {
    DELETED("__DELETED__"),
    UNSET("__UNSET__"),
    ;

    fun code(): String = stateValue

    companion object {
        fun isSpecialStateValue(code: Any?): Boolean =
            if (code is String) {
                code == DELETED.code() || code == UNSET.code()
            } else {
                false
            }

        fun getSpecialStateValue(code: String): SpecialStateValue =
            when (code) {
                DELETED.code() -> DELETED
                UNSET.code() -> UNSET
                else -> throw IllegalArgumentException("Invalid special state value: $code")
            }
    }
}
